### DISCUSSION TITLE
ANN-to-SNN Converter

### DISCUSSION THREAD

**Discussion by cnrgbot** [nengo/nengo-zoo#657] [General Discussion]
> _Archived from the Nengo Discourse forum._ Originally posted by **victkid** on 2020-01-21. (Forum topic #1055, post #1 / id 2753)

Hi,
I'm reading your paper "Training Spiking Deep Networks for Neuromorphic Hardware". It suggests first train the network with ANN, then take the parameters to connect spiking neurons, and continue the training with SNN. Does Nengo currently using the methods mentioned in the paper to train SNN? Does it necessary to first train with ANN? Can I directly train SNN with Nengo, does the performance differs significantly from convert and train? I haven't found this information on Nengo documentation. Please verify. Thank you

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **arvoelke** on 2020-01-22. (Forum topic #1055, post #2 / id 2754)

Hello @​victkid and welcome!

> *Originally @​victkid wrote [post #1, topic #1055]:*
>
> Does Nengo currently using the methods mentioned in the paper to train SNN?

These features are supported by [Nengo-DL](https://www.nengo.ai/nengo-dl/). You can find an example here for [training an SNN via backpropagation](https://www.nengo.ai/nengo-dl/examples/spiking-mnist.html), which references this same paper.

> *Originally @​victkid wrote [post #1, topic #1055]:*
>
> Does it necessary to first train with ANN? Can I directly train SNN with Nengo, does the performance differs significantly from convert and train?

The example trains with an ANN and then converts it into an SNN for inference. This conversion is done automatically for you, and the example provides more details. For a feed-forward network such as this one, the difference in performance (between ANN and SNN) depends on how many neurons there are, how frequently they spike, and the time-constants of the synaptic filters. In particular, a faster time-constant produces a lower latency result with higher variability from spike noise, while a slower time-constant has higher latency but less variance. The expected values of the weighted and filtered spike trains remains the same between the ANN and the SNN (in a statistical sense, due to the pooling across all of the neurons and the averaging across time). These trade-offs receive some mathematical treatment in [Voelker (2019)](https://uwspace.uwaterloo.ca/bitstream/handle/10012/14625/Voelker_Aaron.pdf).

There are also ways you can use Nengo-DL to train directly on spikes in the forward pass, with a custom gradient for the backwards pass for the backpropagation algorithm. This is especially beneficial when training spiking RNNs, since the variance in the spike trains tend to propagate through the recurrent connections. We aim to publish some examples of this kind of training soon. In this context, the dynamics of the synaptic filter also end up playing a useful role that can be exploited computationally (rather than the feed-forward case of trading increased latency for decreased variability), as in the [Nengo examples for building dynamical systems](https://www.nengo.ai/nengo/examples.html#building-dynamical-systems).

**Reply by cnrgbot** [reply]
> _Archived from the Nengo Discourse forum._ Originally posted by **ntouev** on 2022-01-17. (Forum topic #1055, post #3 / id 6237)

Hello @​arvoelke, 

I was reading the question and the answer above since it is exactly was I was looking for, but still I am confused on one thing. The modifications proposed in the paper (removing local contrast normalization, max pooling and applying noise), to be implemented prior to ANN training, are not implemented by the converter but by the user, right?

Thank you

**Comment by cnrgbot** [comment]
> _Archived from the Nengo Discourse forum._ Originally posted by **xchoo** on 2022-01-17. (Forum topic #1055, post #4 / id 6240)

Hi @​ntouev, 

> *Originally @​ntouev wrote [post #3, topic #1055]:*
>
> The modifications proposed in the paper (removing local contrast normalization, max pooling and applying noise), to be implemented prior to ANN training, are not implemented by the converter but by the user, right?
From my knowledge of the ANN-to-SNN converter in NengoDL, yes, you are correct. The techniques proposed in the paper have to be implemented by the user as part of the network's architecture prior to the ANN training process. All the converter does is translate the weights and biases in the ANN to the appropriate weights and biases in the SNN.
